### DISCUSSION TITLE
What do you want to read in a documentation?

### DISCUSSION THREAD

**Discussion by hanspagel** [ueberdosis/tiptap#162] [General Questions]
We'll start to work on a documentation soon. We're pretty thankful for the small community that you build around tiptap, so I thought it would be great to ask what you want to see in a documentation.

I think we can use past issues as a starting point. One of the most asked question is about the styling in the examples.

What questions or what parts of tiptap should we make sure to cover? Thanks for your feedback!

**Comment by Alecyrus** [comment]
Something about collaborative editing?  And can I bind click event to Node/Mark in Node/Mark class? I want to toggle the right click menu when a specific node/mark is clicked even hovered.

**Comment by DanielAlcaraz** [comment]
People are always asking for new extensions. Maybe a full guide to show how to create it and how it works can help to grow up the ecosystem.

**Comment by hanspagel** [comment]
good idea

**Comment by matzeso** [comment]
Hey everyone, I just stumbled across tiptap the other day and really love the idea behind it. I don't know about the requirements of everyone else but for me it is important to decide exactly what kind of extensions a user can use, and also the ability to define custom extensions.

To integrate tiptap with the existing extensions was more or less straightforward so far - but building my own extension takes me way more time than I anticipated. I have already been going through prosemirror documentation and all your example code but I still screw it up.

What I struggled with but thought I now understand is: what is the syntax/options of parseDOM and toDOM, what are generally the available schema properties. What does `group: block` do, what is the difference between `content: text*` or `content: text+`.
How do I need to use commands? 
Especially helpful would be to know what is part of tiptap and what is part of prosemirror, as I have no idea whether I need to look something up in the prosemirror documentation or find out through your examples.

Right now, for example, I just want to have a button insert some specific HTML format (in this case a button, but it is wrapped in a table 

```<table><tr><td><a href="scrumpy.com"><button style="......">this is the text I then edit after I clicked</button></a></td></tr></table>```

I thought I was going in the right direction but my editor doesn't seem to parse anything and as soon as I edit the text it all gets replaced by a standard paragraph again. 

Long story short: Creating your own extensions, step by step, the different ways to use them (insert, replace), etc. would be really helpful.
I saw you guys are sitting in Berlin as well, would be nice to have a coffee sometime and chat about it if you like.

Bests,
Matthias

**Comment by Alecyrus** [comment]
@matzeso  
## How to create an extension?
[These](https://github.com/scrumpy/tiptap/tree/master/packages/tiptap-extensions/src) are good examples.   Like [Underline.js](https://github.com/scrumpy/tiptap/blob/master/packages/tiptap-extensions/src/marks/Bold.js), you could learn how to define a new mark, and you could rewrite the `Underline.js` to implement more complex underline, like underline with color and line style (wavy? solid? dotted?)
There is some code about `Underline` with styles in my project.

![image](https://user-images.githubusercontent.com/13865581/51584593-06c52800-1f11-11e9-8911-793632865565.png)

```javascript
 get schema() {
    return {
      attrs: {
        color: {
          default: 1,
        },
        lineStyle: {
          default: 1,
        },
      },
      parseDOM: [
        {
          tag: 'u',
        },
        {
          style: 'text-decoration',
          getAttrs: value => value === 'underline',
        },
      ],
      toDOM(mark) {
        return ['u', {
          class: `underline-${hlcolor.color[mark.attrs.color]}`,
          style: `text-decoration-style: ${lineStyle[mark.attrs.lineStyle]};text-decoration-color:${hlcolor.map[hlcolor.color[mark.attrs.color]]}`,
        }, 0];
      },
    };
  }
```


In fact, it is to define new nodes , marks or plugins.   You could learn these concepts from [Prosemirror Documentation](http://prosemirror.net/docs/guide/).  That will help you to understand how Prosemirror and tiptap works. 

Documentation have explained all the things.

#### what is the syntax/options of parseDOM and toDOM, what are generally the available schema properties
Check [this](http://prosemirror.net/docs/ref/)


####  what is the difference between content: text* or content: text+.
> You can say, for example "paragraph" for “one paragraph”, or "paragraph+" to express “one or more paragraphs”. Similarly, "paragraph*" means “zero or more paragraphs” and "caption?" means “zero or one caption node”. You can also use regular-expression-like ranges, such as {2} (“exactly two”) {1, 5} (“one to five”) or {2,} (“two or more”) after node names. 

#### What does `group: block` do
>Some groups of element types will appear multiple types in your schema—for example you might have a concept of “block” nodes, that may appear at the top level but also nested inside of blockquotes. You can create a node group by giving your node specs a group property, and then refer to that group by its name in your expressions.
```javascript
const groupSchema = new Schema({
  nodes: {
    doc: {content: "block+"},
    paragraph: {group: "block", content: "text*"},
    blockquote: {group: "block", content: "block+"},
    text: {}
  }
})
```

**Comment by stargazing-dino** [comment]
I'm trying to implement Named Entity Recognition highlighting inside an editor and though I've mostly successfully followed @Alecyrus instructions in copying and learning from the extensions already here, I felt completely overwhelmed moving from the relatively simple Underline.js example to the Suggestions.js one. And I wasn't even sure if a _plugin_ was what I needed in order to implement NER highlighting. I only assumed because it has access to state.

I guess my suggestion for the documentation would be to have more varied examples, thus easier for some to find the right tool for what they need faster. I think the [examples here](http://prosemirror.net/examples/) would do a lot of good if the docs were to mention them and have implementations using tiptap.

Love the library btw

**Comment by Alecyrus** [comment]
@Nolence  Thank you for your feedback. If you want to implement NER highlighting, I think the easiest way is to implement a plugin that could scan the text in all `text` nodes, and add specific `mark` that you expect to certain text in `text` nodes. But it seems like not an elegant way. Of course, you have the access to state in `plugin`, see [this](http://prosemirror.net/docs/ref/#state.PluginSpec.view),  the next state and pre state can be visited through:
> some code from [Suggestions.js](https://github.com/scrumpy/tiptap/blob/master/packages/tiptap-extensions/src/plugins/Suggestions.js)
``` javascript

view() {
      return {
        update: (view, prevState) => {
          const prev = this.key.getState(prevState)
          const next = this.key.getState(view.state)
          ...
```
I believe the [documentation](https://github.com/scrumpy/tiptap-docs) will be coming soon.

**Comment by stargazing-dino** [comment]
Thanks @Alecyrus for your help. Right now I'm stuck getting the asynchronous part working. I'm firing off a transaction with metadata inside a debounced function in the plugin's `view.update` and catching it in the plugin's `state.apply` method, but weirdly `tr.docChanged` and `tr.getMeta('makeAPIRequest)` are never true at the same time for me to create a decoration set. I'll [continue my question](https://discuss.prosemirror.net/t/async-requests-in-linter-example/1772) along with the actual code at the prose-mirror discussion board though.

**TL;DR**
To keep on topic, I recently stumbled on a video [going through Draft-js](https://www.youtube.com/watch?v=vOZAO3jFSHI) and couldn't help but share the [interactive website](https://draft-js-samples.now.sh/Home) created for the demonstration. It would be awesome if something like that could be worked out to show not only `tiptap`'s API but also the fundamentals of how `prose-mirror` works.

**Comment by erickwilder** [comment]
Thank you @hanspagel for opening this topic for discussion. What I've seen happening in our team is that sometimes it's difficult for someone to tell if something in the interfaces exposed by tiptap is from Prosemirror _or_ from tiptap itself. Then we get stuck trying to search for answers in the wrong place. A very good example is here, mentioned by @matzeso - all these constructs with `text*`, `block+` et al are _from Prosemirror_ but there is close to nothing about guidance to the relevant documentation (or forums, stack overflow, etc) in the Prosemirror realm.

Another (small) issue is that the README file grew quite a bit and perhaps adopting a more well-structured way to write docs (like GitBook and similar options) would make _discoverability_ and consistency key to a good understanding of the project and its underlying components.

**Comment by hanspagel** [comment]
Thanks, Erick!

There is a repo (based on VuePress) already. It's deploying to https://tiptap.scrumpy.io/docs

We didn't start with the actual work of writing, but if someone wants to add a few things we're happy to accept pull requests. 

 https://github.com/scrumpy/tiptap-docs

**Comment by erickwilder** [comment]
@hanspagel, I just submitted [PR#1](https://github.com/scrumpy/tiptap-docs/pull/1) on `tiptap-docs` repo. 🎉

**Comment by kvanska** [comment]
Hello there, 

I'm posting this question here in the open discussion, maybe something that could be included in the documentation. 

I'm trying to change the Mention example to allow calling REST api for getting the names (so I don't have to load tens of thousands of names ...). This is kind of basic use-case for this feature. I just don't get where to write the async axios method to filter the names. I tried overwriting onFilter but this can't be async .. 

Am I doing something totally wrong here?

thanks for great work!

**Comment by rasliche** [comment]
I'm so glad I found this issue today. Tiptap looks amazing, and I a really building from the ground up and learning as I go. I will do my best to help out with some documentation as I figure it out. I'm just happy to see that this project is not "dead" by any means. Where else does the community gather? Is there a repository (not necessarily a literal github repo) of extensions for tiptap anywhere?

**Comment by hanspagel** [comment]
Erick added list of built-in extensions here: https://tiptap.scrumpy.io/docs/extensions/built-in.html (Work in Progress)

I don't know about community built extensions yet, but sounds like we should put up a collection somewhere to let people add theirs.

**Comment by Alecyrus** [comment]
Consider to create a repo `awesome-tiptap` for 3-rd party extensions.

**Comment by Hyperblaster** [comment]
Wooo, Been searching around for WYSIWYG editors for app platform and so glad i found tiptap.
This is EXACTLY what i've been hoping and dreaming about for a while.

Keen to contribute and to help with writing documentation.

First thing for me would be filling out the documentation on the extensions, and like someone else mentioned above, making @mentions asynchronous

**Comment by marcusamargi** [comment]
@kvanska @Hyperblaster  I believe I have found a way to populate mentions from remote data. 

1) Make a local copy of the Suggestions.js plugin.
2) Change the plugin's view's update function to be `async`.
3) Inside the function, `await` the return of `onFilter` when it gets assigned to `items` in the `props`.
4) Make a local copy of the Mention.js node, and modify to import your local Suggestions.js plugin.
5) In your component that uses the editor, add your local Mention plugin.
6) For `items` you can simply do `items: () => []`
7) For `onFilter` make that function `async` and `await` your api call within in. Be sure to return an array of objects composed of `id` and `name` keys.

**Comment by hyvyys** [comment]
> #### what is the syntax/options of parseDOM and toDOM, what are generally the available schema properties
> 
> Check [this](http://prosemirror.net/docs/ref/)

A bit unclear... I suspect it is somewhere in there. But where? The naming conventions are different.

A general remark: in the stubs in the docs, instead of the TBD notes, there could be links to specific sections of ProseMirror docs.

**Comment by jonasdumas** [comment]
Would be so excellent to have:
- drag n drop block example (paragraphs, h1, h2, etc)
- a list with all extensions made made from the community for tiptap

**Comment by UpsylonDev** [comment]
hello there ;-)  tiptap is really great thanks.
a possible improvement in the doc is the way to add content to the content in a programmed way (for example adding text from the clipboard via a button?)
this.editor.content = "<data>" dont work? ;-)

... and also having to read 2 ProseMirror / tiptap documents to understand is really ... strange and confusing I find: - /

**Comment by hanspagel** [comment]
Thanks for all your suggestions! Just wanted to let you know that we finally started to write a proper documentation for tiptap 2. 🙂 

If you want to support us, please sponsor our work: [https://github.com/sponsors/ueberdosis](https://github.com/sponsors/ueberdosis) Every single sponsorship is a huge motivation to improve tiptap for everyone. 💖

**Comment by 1421gcs** [comment]
Since #160, Six years have passed, and we still cannot easily reproduce any of the examples in the documentation. Although you have provided the complete component code, the lack of CSS code makes it difficult to reproduce. Additionally, I noticed that all the examples on the official website ultimately point to https://embed.tiptap.dev , but I cannot find the source code for this website itself. All of these obstacles have hindered my ability to quickly understand the Tiptap editor!!!

**Reply by bdbch** [reply]
It takes one click to find the demos on embed.tiptap.dev here with all code examples included:
https://github.com/ueberdosis/tiptap/tree/develop/demos

If you want a quick start including styles, try our simple editor template.
https://tiptap.dev/docs/ui-components/templates/simple-editor

**Reply by 1421gcs** [reply]
Thank you! But I hope you can include it in the documentation :)
Once again, thank you to the tiptap team for their hard work.

**Reply by bdbch** [reply]
The demos are included in every page and are usually not meant to be straightup visible. If you are interested in their code you can always look in the repo. No need to additionally add them to the documentation.

The templates are prominently mentioned on the Docs landing page.

**Comment by bdbch** [comment]
Hey everyone! Since this discussion is quite old and things have changed a lot since it started, we're going to close it now. If you have questions or suggestions about the documentation, please check out the current Discussions or, even better, open an issue directly in the Tiptap Docs repo. Thanks for all your input and for helping make Tiptap better!
